### ISSUE TITLE
Twig_Loader_String is deprecated

### ISSUE THREAD

**Issue by CarsonF** [bolt/bolt#3424]
Safe twig uses `Twig_Loader_String` which is deprecated as of `1.18.1`

**Comment by rarila** [comment]
Some hint to solve this: https://github.com/twigphp/Twig/pull/1641

**Comment by bobdenotter** [comment]
Ah, that workaround seems doable.. Great example, too! 

```php
$env = new \Twig_Environment(new \Twig_Loader_Array());
$template = $env->createTemplate('Hello, {{ name }}');
echo $env->render($template, array('name' => 'Bob')); // Hello, Bob
```

**Comment by rarila** [comment]
@bobdenotter you only like it because of that example name ;-)

**Comment by jeroenvdgulik** [comment]
```php
$env = new \Twig_Environment(new \Twig_Loader_Array());
```

This won't work as `Twig_Loader_Array()` expects a template as constructor parameter

**Comment by GawainLynch** [comment]
@jeroenvdgulik yeah, we're just planning on implementing our own class sans a better idea.

**Comment by jeroenvdgulik** [comment]
It took me some time to figure this out so maybe this is useful for you guys. This works:

```php
$env = new \Twig_Environment(new \Twig_Loader_Array(['my_template_name' => 'Hello, {{ name }}']));
$html = $env->render('my_template_name', array('name' => 'Bob'));
```

**Comment by CarsonF** [comment]
I think what we need to do is not have a separate `safe_twig` environment, but just use the sandbox extension they provide.

**Comment by GawainLynch** [comment]
Bumping this to a 2.4 milestone as @CarsonF has a branch in progress, but has too much flux for this stage in 2.3-dev.

**Comment by SahAssar** [comment]
Any progress on this?

**Comment by CarsonF** [comment]
Nope. We could probably just copy their source. 
But there's a reason they deprecated it. It's basically impossible to cache.

**Comment by GawainLynch** [comment]
Ha… I've been meaning to bring this up with you for a couple of weeks, @CarsonF 

It isn't super urgent right now, as it won't bite until next major anyway. But keep this open to track @SahAssar

**Comment by CarsonF** [comment]
Fixed in 3.3 with sandbox
